File: docs/algorithms/code/algorithms/sortedArrayToBST.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function sortedArrayToBSTHelper(
  nums: number[],
  left: number,
  right: number
): TreeNode | null {
  if (left > right) {
    return null;
  }

  const mid = (left + right) >> 1;

  const root = new TreeNode(nums[mid]);

  root.left = sortedArrayToBSTHelper(nums, left, mid - 1);
  root.right = sortedArrayToBSTHelper(nums, mid + 1, right);

  return root;
}

function sortedArrayToBST(nums: number[]): TreeNode | null {
  return sortedArrayToBSTHelper(nums, 0, nums.length - 1);
}

File: docs/algorithms/code/algorithms/sortedArrayToBST.go
// 思路：每次找出数组的中间元素作为根节点，然后递归地构造左子树和右子树。
// 复杂度分析：时间复杂度为 O(n)， 空间复杂度 O(n)
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sortedArrayToBST(nums []int) *TreeNode {
	if len(nums) == 0 {
		return nil
	}

	mid := len(nums) / 2

	root := &TreeNode{Val: nums[mid]}
	root.Left = sortedArrayToBST(nums[:mid])
	root.Right = sortedArrayToBST(nums[mid+1:])
	return root
}